modules/heat_conduction/src/userobjects/ViewFactorRayStudy.C

```diff
@@ -0,0 +1,554 @@
+  // No need to use Ray registration
+  params.set<bool>("_use_ray_registration") = false;
+  // Do not need to bank Rays on completion
+  params.set<bool>("_bank_rays_on_completion") = false;
+
+  params.addClassDescription(
+      "This ray study is used to compute view factors in cavities with obstruction. It sends out "
+      "rays from surfaces bounding the radiation cavity into a set of directions determined by an "
+      "angular quadrature. The rays are tracked and view factors are computed by determining the "
+      "surface where the ray dies.");
+  return params;
+}
+
+ViewFactorRayStudy::ViewFactorRayStudy(const InputParameters & parameters)
+  : RayTracingStudy(parameters),
+    _bnd_ids_vec(_mesh.getBoundaryIDs(getParam<std::vector<BoundaryName>>("boundary"))),
+    _bnd_ids(_bnd_ids_vec.begin(), _bnd_ids_vec.end()),
+    _internal_convention(getParam<MooseEnum>("internal_convention")),
+    _ray_index_start_bnd_id(registerRayAuxData("start_bnd_id")),
+    _ray_index_start_total_weight(registerRayAuxData("start_total_weight")),
+    _fe_face(FEBase::build(_mesh.dimension(), FEType(CONSTANT, MONOMIAL))),
+    _q_face(QBase::build(Moose::stringToEnum<QuadratureType>(getParam<MooseEnum>("face_type")),
+                         _mesh.dimension() - 1,
+                         Moose::stringToEnum<Order>(getParam<MooseEnum>("face_order")))),
+    _is_3d(_mesh.dimension() == 3),
+    _threaded_vf_info(libMesh::n_threads())
+{
+  _fe_face->attach_quadrature_rule(_q_face.get());
+  _fe_face->get_xyz();
+
+  // create angular quadrature
+  if (!_is_3d)
+  {
+    // In 2D, we integrate over angle theta instead of mu = cos(theta)
+    // The integral over theta is approximated using a Gauss Legendre
+    // quadrature. The integral we need to approximate is given by:
+    //
+    // int_{-pi/2}^{pi/2} cos(theta) d theta
+    //
+    // We get abscissae x and weight w for range of integration
+    // from 0 to 1 and then rescale it to the integration range
+    //
+    std::vector<Real> x;
+    std::vector<Real> w;
+    RayTracingAngularQuadrature::gaussLegendre(
+        2 * getParam<unsigned int>("polar_quad_order"), x, w);
+
+    _2d_aq_angles.resize(x.size());
+    _2d_aq_weights.resize(x.size());
+    for (unsigned int j = 0; j < x.size(); ++j)
+    {
+      _2d_aq_angles[j] = (2 * x[j] - 1) * M_PI / 2;
+      _2d_aq_weights[j] = w[j] * M_PI;
+    }
+    _num_dir = _2d_aq_angles.size();
+  }
+  else
+  {
+    _3d_aq = libmesh_make_unique<RayTracingAngularQuadrature>(
+        _mesh.dimension(),
+        getParam<unsigned int>("polar_quad_order"),
+        4 * getParam<unsigned int>("azimuthal_quad_order"),
+        /* mu_min = */ 0,
+        /* mu_max = */ 1);
+
+    _num_dir = _3d_aq->numDirections();
+  }
+}
+
+void
+ViewFactorRayStudy::initialSetup()
+{
+  RayTracingStudy::initialSetup();
+
+  // We optimized away RayKernels, so don't allow them
+  if (hasRayKernels(/* tid = */ 0))
+    mooseError("Not compatible with RayKernels.");
```

Comment:
Add details, could be user facing

Reply:
unless new error logs are up

Reply:
New errors will say something like

The following error occurred in the object "ViewFactorRayStudy" named "blah".....

Not compatible with RayKernels

Reply:
ok that's fine. New errors are great